brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1316 @@
+        # between voxels. This means that V is a diagonal matrix instead of
+        # an identity matrix. The parameter includes the lower triangular
+        # part of the cholesky decomposition
+        # of U (flattened), then tan(rho1*pi/2) where rho1 is
+        # each voxel's autoregressive coefficient (assumging AR(1) model).
+        # Such parametrization avoids the need of boundaries
+        # for parameters.
+
+        LL = 0.0  # log likelihood
+
+        n_l = np.size(l_idx[0])
+        # the number of parameters in the index of lower-triangular matrix
+        # This indexing allows for parametrizing only
+        # part of the lower triangular matrix (non-full rank covariance matrix)
+        L = np.zeros([n_C, rank])
+        # lower triagular matrix L, cholesky decomposition of U
+        L[l_idx] = param[0:n_l]
+
+        a1 = param[n_l:n_l + n_V]
+        rho1 = 2.0 / np.pi * np.arctan(a1)  # auto-regressive coefficients
+
+        SNR2 = np.exp(log_SNR2)
+        # each element of SNR2 is the ratio of the diagonal element on V
+        # to the variance of the fresh noise in that voxel
+
+        # derivatives
+        deriv_L = np.zeros(np.shape(L))
+        deriv_a1 = np.zeros(np.shape(rho1))
+
+        YTAY = YTY_diag - rho1 * YTDY_diag + rho1**2 * YTFY_diag
+        # dimension: space,
+        XTAX = XTX[np.newaxis, :, :] - rho1[:, np.newaxis, np.newaxis]\
+            * XTDX[np.newaxis, :, :] + rho1[:, np.newaxis, np.newaxis]**2\
+            * XTFX[np.newaxis, :, :]
+        # dimension: space*feature*feature
+        XTAY = XTY - rho1 * XTDY + rho1**2 * XTFY
+        # dimension: feature*space
+        LTXTAY = np.dot(L.T, XTAY)
+        # dimension: rank*space
+        LAMBDA_i = np.zeros([n_V, rank, rank])
+        for i_v in range(n_V):
+            LAMBDA_i[i_v, :, :] = np.eye(rank) \
+                + np.dot(np.dot(L.T, XTAX[i_v, :, :]), L) * SNR2[i_v]
+        # dimension: space*rank*rank
+        LAMBDA = np.linalg.inv(LAMBDA_i)
+        # dimension: space*rank*rank
+        YTAXL_LAMBDA = np.einsum('ijk,ki->ij', LAMBDA, LTXTAY)
+        # dimension: space*rank
+        YTAXL_LAMBDA_LT = np.dot(YTAXL_LAMBDA, L.T)
+        # dimension: space*feature (feature can be larger than rank)
+
+        sigma2 = (YTAY - SNR2 * np.sum(YTAXL_LAMBDA_LT * XTAY.T, axis=1)) \
+            / n_T
+        # dimension: space,
+
+        LL = -np.sum(np.log(sigma2)) * n_T * 0.5 \
+            + np.sum(np.log(1 - rho1**2)) * 0.5 \
+            - np.sum(np.log(np.linalg.det(LAMBDA_i))) * 0.5 \
+            - n_T / 2.0
+        XTAXL = np.dot(XTAX, L)
+        # dimension: space*feature*rank
+        deriv_L = -np.einsum('ijk,ikl,i', XTAXL, LAMBDA, SNR2) - \
+            np.einsum('ijk,ik,il,i', XTAXL, YTAXL_LAMBDA, YTAXL_LAMBDA,
+                      SNR2**2 / sigma2) \
+            + np.dot(XTAY / sigma2 * SNR2, YTAXL_LAMBDA)
+        # dimension: feature*rank
+        tmp = -XTDX + 2 * rho1[:, np.newaxis, np.newaxis] * XTFX
```

Comment:
I renamed it as dXTAX_drho1 now - the derivative of the product X'AX wrt to the AR(1) coefficient, where X'AX is explained somewhere above. A/sigma2 is the inverse covariance matrix of noise in each voxel.